File: sdk/python/forge_sdk/worker.py
"""Forge Python Worker — connects to Coordinator, executes tasks."""

import json
import logging
import signal
import socket
import threading
import uuid
from concurrent import futures
from typing import Callable, Dict, Optional

import grpc
from google.protobuf import timestamp_pb2

from .decorators import TaskContext, get_registered_handlers
from .generated import worker_pb2, worker_pb2_grpc

logger = logging.getLogger("forge.worker")


class _WorkerServicer(worker_pb2_grpc.WorkerServiceServicer):
    """gRPC servicer implementing WorkerService.

    Handles ExecuteTask and Heartbeat RPCs initiated by the Coordinator.
    """

    def __init__(self, worker: "Worker") -> None:
        self._worker = worker

    def Register(self, request, context):
        """No-op — the Worker registers itself with the Coordinator, not the other way around."""
        return worker_pb2.RegisterResponse(accepted=True)

    def Heartbeat(self, request_iterator, context):
        """Bidirectional heartbeat stream.

        The Coordinator sends HeartbeatPing messages; we respond with HeartbeatPong
        containing current active task count and capacity.
        """
        for ping in request_iterator:
            with self._worker._active_lock:
                active = self._worker._active_tasks

            now = timestamp_pb2.Timestamp()
            now.GetCurrentTime()
            yield worker_pb2.HeartbeatPong(
                worker_id=self._worker._worker_id,
                active_tasks=active,
                capacity=self._worker._capacity,
                timestamp=now,
            )

    def ExecuteTask(self, request: worker_pb2.TaskRequest, context) -> worker_pb2.TaskResponse:
        """Dispatch an incoming task to the appropriate handler."""
        handler_name = request.handler
        handler_fn = self._worker._handlers.get(handler_name)

        if handler_fn is None:
            return worker_pb2.TaskResponse(
                task_id=request.task_id,
                success=False,
                error_msg=f"handler '{handler_name}' not found",
            )

        # Build TaskContext
        ctx = TaskContext(
            task_id=request.task_id,
            workflow_id=request.workflow_id,
            task_name=request.task_name,
            handler=handler_name,
            timeout_ms=request.timeout_ms,
            worker_id=self._worker._worker_id,
        )

        # Decode JSON input
        try:
            params = json.loads(request.input) if request.input else {}
        except (json.JSONDecodeError, UnicodeDecodeError) as exc:
            return worker_pb2.TaskResponse(
                task_id=request.task_id,
                success=False,
                error_msg=f"failed to decode input JSON: {exc}",
            )

        # Track active tasks
        with self._worker._active_lock:
            self._worker._active_tasks += 1

        try:
            result = handler_fn(ctx, params)
            output_bytes = json.dumps(result).encode("utf-8") if result is not None else b"{}"
            return worker_pb2.TaskResponse(
                task_id=request.task_id,
                success=True,
                output=output_bytes,
            )
        except Exception as exc:
            logger.error("handler '%s' failed for task %s: %s", handler_name, request.task_id, exc)
            return worker_pb2.TaskResponse(
                task_id=request.task_id,
                success=False,
                error_msg=str(exc),
            )
        finally:
            with self._worker._active_lock:
                self._worker._active_tasks -= 1


class Worker:
    """Forge Python Worker.

    Connects to a Coordinator via gRPC, auto-registers with labels/handlers/capacity,
    serves a gRPC endpoint for Heartbeat and ExecuteTask RPCs from the Coordinator,
    and supports graceful shutdown.

    Args:
        coordinator: Coordinator gRPC address (e.g., ``"localhost:8080"``).
        worker_id: Unique worker identifier. Auto-generated UUID if omitted.
        capacity: Maximum concurrent tasks this Worker can handle.
        labels: Key-value scheduling hints (e.g., ``{"gpu": "true"}``).
        listen_port: Port for the Worker's gRPC server (0 = auto-pick).
    """

    def __init__(
        self,
        coordinator: str,
        worker_id: Optional[str] = None,
        capacity: int = 5,
        labels: Optional[Dict[str, str]] = None,
        listen_port: int = 0,
    ) -> None:
        self._coordinator_addr = coordinator
        self._worker_id = worker_id or str(uuid.uuid4())
        self._capacity = capacity
        self._labels = labels or {}
        self._listen_port = listen_port

        # Handler dispatch table: populated from decorator registry + manual registration
        self._handlers: Dict[str, Callable] = {}

        # Active task tracking
        self._active_tasks = 0
        self._active_lock = threading.Lock()

        # Shutdown coordination
        self._stop_event = threading.Event()
        self._server: Optional[grpc.Server] = None
        self._channel: Optional[grpc.Channel] = None

    @property
    def worker_id(self) -> str:
        """Return this Worker's unique identifier."""
        return self._worker_id

    @property
    def handlers(self) -> Dict[str, Callable]:
        """Return the current handler dispatch table."""
        return dict(self._handlers)

    def register_handler(self, name: str, func: Callable) -> None:
        """Manually register a task handler.

        Args:
            name: Handler identifier (e.g., ``"ai.generate"``).
            func: Callable accepting ``(ctx: TaskContext, params: dict) -> dict``.
        """
        self._handlers[name] = func

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def start(self) -> None:
        """Start the Worker (blocking).

        1. Collects handlers from the decorator registry.
        2. Starts a gRPC server for receiving ExecuteTask and Heartbeat RPCs.
        3. Registers with the Coordinator.
        4. Blocks until ``stop()`` is called or a signal is received.
        """
        # Merge decorator-registered handlers
        for name, fn in get_registered_handlers().items():
            if name not in self._handlers:
                self._handlers[name] = fn

        if not self._handlers:
            raise RuntimeError("no task handlers registered — use @task_handler or register_handler()")

        # Install signal handlers for graceful shutdown (only works in main thread)
        import threading as _threading
        if _threading.current_thread() is _threading.main_thread():
            signal.signal(signal.SIGINT, self._signal_handler)
            signal.signal(signal.SIGTERM, self._signal_handler)

        # Start gRPC server
        self._server = grpc.server(futures.ThreadPoolExecutor(max_workers=self._capacity))
        servicer = _WorkerServicer(self)
        worker_pb2_grpc.add_WorkerServiceServicer_to_server(servicer, self._server)

        if self._listen_port == 0:
            self._listen_port = self._server.add_insecure_port("[::]:0")
        else:
            self._server.add_insecure_port(f"[::]:{self._listen_port}")

        self._server.start()
        logger.info(
            "worker %s listening on port %d (capacity=%d, handlers=%s)",
            self._worker_id,
            self._listen_port,
            self._capacity,
            list(self._handlers.keys()),
        )

        # Register with Coordinator
        self._register()

        # Block until stop
        self._stop_event.wait()
        self._shutdown()

    def stop(self) -> None:
        """Signal the Worker to shut down gracefully."""
        self._stop_event.set()

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _signal_handler(self, signum, frame) -> None:
        logger.info("received signal %d, shutting down", signum)
        self.stop()

    def _get_worker_addr(self) -> str:
        """Return the advertised gRPC address of this Worker."""
        hostname = socket.gethostname()
        return f"{hostname}:{self._listen_port}"

    def _register(self) -> None:
        """Register this Worker with the Coordinator via gRPC."""
        self._channel = grpc.insecure_channel(self._coordinator_addr)
        stub = worker_pb2_grpc.WorkerServiceStub(self._channel)

        registration = worker_pb2.WorkerRegistration(
            id=self._worker_id,
            addr=self._get_worker_addr(),
            labels=self._labels,
            capacity=self._capacity,
            handlers=list(self._handlers.keys()),
        )
        request = worker_pb2.RegisterRequest(registration=registration)

        try:
            response = stub.Register(request)
            if response.accepted:
                logger.info("registered with coordinator: %s", response.message)
            else:
                raise RuntimeError(f"coordinator rejected registration: {response.message}")
        except grpc.RpcError as exc:
            raise ConnectionError(
                f"failed to register with coordinator at {self._coordinator_addr}: {exc}"
            ) from exc

    def _shutdown(self) -> None:
        """Perform graceful shutdown."""
        logger.info("shutting down worker %s", self._worker_id)

        # Stop gRPC server with grace period for in-flight tasks
        if self._server is not None:
            self._server.stop(grace=30)

        # Close channel to coordinator
        if self._channel is not None:
            self._channel.close()

        logger.info("worker %s stopped", self._worker_id)

File: sdk/cpp/src/worker.cpp
// Copyright 2024 Forge Authors. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

#include <forge/worker.h>

#include <algorithm>
#include <atomic>
#include <chrono>
#include <condition_variable>
#include <csignal>
#include <iostream>
#include <map>
#include <mutex>
#include <random>
#include <string>
#include <thread>
#include <vector>

#include <grpcpp/grpcpp.h>
#include <nlohmann/json.hpp>

#include <google/protobuf/timestamp.pb.h>
#include <google/protobuf/util/time_util.h>

#include "grpc_client.h"
#include "worker.grpc.pb.h"
#include "worker.pb.h"

namespace forge {

namespace forgev1 = ::forge::v1;

// ---------------------------------------------------------------------------
// Generate a simple UUID-v4-like identifier.
// ---------------------------------------------------------------------------
static std::string generate_uuid() {
    static std::mt19937 rng(std::random_device{}());
    static std::uniform_int_distribution<int> dist(0, 15);
    static const char* hex = "0123456789abcdef";

    std::string uuid = "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx";
    for (auto& c : uuid) {
        if (c == 'x') {
            c = hex[dist(rng)];
        } else if (c == 'y') {
            c = hex[(dist(rng) & 0x3) | 0x8];
        }
    }
    return uuid;
}

// ---------------------------------------------------------------------------
// WorkerServicer — gRPC server-side implementation of WorkerService.
// ---------------------------------------------------------------------------
class WorkerServicer final : public forgev1::WorkerService::Service {
public:
    WorkerServicer(
        const std::string& worker_id,
        std::map<std::string, std::unique_ptr<TaskHandler>>& handlers,
        std::atomic<int32_t>& active_tasks,
        int capacity)
        : worker_id_(worker_id),
          handlers_(handlers),
          active_tasks_(active_tasks),
          capacity_(capacity) {}

    grpc::Status Register(
        grpc::ServerContext* /*context*/,
        const forgev1::RegisterRequest* /*request*/,
        forgev1::RegisterResponse* response) override {
        response->set_accepted(true);
        return grpc::Status::OK;
    }

    grpc::Status Heartbeat(
        grpc::ServerContext* /*context*/,
        grpc::ServerReaderWriter<forgev1::HeartbeatPong,
                                 forgev1::HeartbeatPing>* stream) override {
        std::cerr << "[INFO] heartbeat stream opened for worker "
                  << worker_id_ << "\n";

        forgev1::HeartbeatPing ping;
        while (stream->Read(&ping)) {
            forgev1::HeartbeatPong pong;
            pong.set_worker_id(worker_id_);
            pong.set_active_tasks(active_tasks_.load());
            pong.set_capacity(capacity_);

            auto* ts = pong.mutable_timestamp();
            auto now = std::chrono::system_clock::now();
            auto epoch = now.time_since_epoch();
            auto seconds = std::chrono::duration_cast<std::chrono::seconds>(epoch);
            auto nanos = std::chrono::duration_cast<std::chrono::nanoseconds>(
                             epoch - seconds)
                             .count();
            ts->set_seconds(seconds.count());
            ts->set_nanos(static_cast<int32_t>(nanos));

            if (!stream->Write(pong)) {
                std::cerr << "[ERROR] failed to send heartbeat pong for worker "
                          << worker_id_ << "\n";
                break;
            }
        }

        std::cerr << "[INFO] heartbeat stream closed for worker "
                  << worker_id_ << "\n";
        return grpc::Status::OK;
    }

    grpc::Status ExecuteTask(
        grpc::ServerContext* /*context*/,
        const forgev1::TaskRequest* request,
        forgev1::TaskResponse* response) override {

        active_tasks_.fetch_add(1);

        response->set_task_id(request->task_id());

        auto it = handlers_.find(request->handler());
        if (it == handlers_.end()) {
            active_tasks_.fetch_sub(1);
            response->set_success(false);
            response->set_error_msg(
                "unknown handler: " + request->handler());
            return grpc::Status::OK;
        }

        // Parse JSON input parameters.
        nlohmann::json params;
        if (!request->input().empty()) {
            try {
                params = nlohmann::json::parse(request->input());
            } catch (const nlohmann::json::exception& e) {
                active_tasks_.fetch_sub(1);
                response->set_success(false);
                response->set_error_msg(
                    std::string("failed to decode input JSON: ") + e.what());
                return grpc::Status::OK;
            }
        }

        // Build TaskContext.
        TaskContext ctx(
            request->task_id(),
            request->workflow_id(),
            request->task_name(),
            request->handler(),
            request->timeout_ms(),
            worker_id_,
            std::move(params));

        // Execute handler.
        try {
            TaskResult result = it->second->execute(ctx);
            response->set_success(result.success());

            if (result.success()) {
                std::string output_json = result.output().dump();
                response->set_output(output_json);
            } else {
                response->set_error_msg(result.error_message());
            }
        } catch (const std::exception& e) {
            response->set_success(false);
            response->set_error_msg(
                std::string("handler exception: ") + e.what());
        }

        active_tasks_.fetch_sub(1);
        return grpc::Status::OK;
    }

private:
    std::string worker_id_;
    std::map<std::string, std::unique_ptr<TaskHandler>>& handlers_;
    std::atomic<int32_t>& active_tasks_;
    int capacity_;
};

// ---------------------------------------------------------------------------
// Worker::Impl — PIMPL implementation.
// ---------------------------------------------------------------------------
class Worker::Impl {
public:
    Impl(const std::string& coordinator_addr, const std::string& worker_id)
        : coordinator_addr_(coordinator_addr),
          worker_id_(worker_id.empty() ? generate_uuid() : worker_id),
          capacity_(5),
          listen_port_(0),
          active_tasks_(0),
          running_(false) {}

    ~Impl() {
        stop();
    }

    void register_handler(std::unique_ptr<TaskHandler> handler) {
        const std::string handler_name = handler->name();
        handlers_[handler_name] = std::move(handler);
    }

    void set_capacity(int capacity) { capacity_ = capacity; }

    void set_labels(const std::map<std::string, std::string>& labels) {
        labels_ = labels;
    }

    void set_listen_port(int port) { listen_port_ = port; }

    const std::string& id() const noexcept { return worker_id_; }

    void start() {
        if (handlers_.empty()) {
            throw std::runtime_error(
                "no task handlers registered — call register_handler() before start()");
        }

        running_.store(true);

        // Start gRPC server.
        start_grpc_server();

        // Register with Coordinator.
        register_with_coordinator();

        // Block until stop() is called.
        {
            std::unique_lock<std::mutex> lock(mu_);
            cv_.wait(lock, [this] { return !running_.load(); });
        }

        shutdown();
    }

    void stop() {
        bool expected = true;
        if (running_.compare_exchange_strong(expected, false)) {
            cv_.notify_all();
        }
    }

private:
    void start_grpc_server() {
        std::string server_addr =
            "0.0.0.0:" + std::to_string(listen_port_);

        grpc::ServerBuilder builder;
        int selected_port = 0;
        builder.AddListeningPort(
            server_addr, grpc::InsecureServerCredentials(), &selected_port);

        servicer_ = std::make_unique<WorkerServicer>(
            worker_id_, handlers_, active_tasks_, capacity_);
        builder.RegisterService(servicer_.get());

        server_ = builder.BuildAndStart();
        if (!server_) {
            throw std::runtime_error("failed to start gRPC server on " + server_addr);
        }

        listen_port_ = selected_port;
        std::cerr << "[INFO] worker " << worker_id_
                  << " listening on port " << listen_port_
                  << " (capacity=" << capacity_ << ", handlers=[";
        bool first = true;
        for (const auto& [name, _] : handlers_) {
            if (!first) std::cerr << ", ";
            std::cerr << name;
            first = false;
        }
        std::cerr << "])\n";
    }

    void register_with_coordinator() {
        internal::GrpcClient client(coordinator_addr_);

        std::vector<std::string> handler_names;
        handler_names.reserve(handlers_.size());
        for (const auto& [name, _] : handlers_) {
            handler_names.push_back(name);
        }

        std::string worker_addr = get_worker_addr();

        if (!client.register_worker(
                worker_id_, worker_addr, labels_, capacity_, handler_names)) {
            throw std::runtime_error(
                "failed to register with coordinator at " + coordinator_addr_);
        }
    }

    std::string get_worker_addr() const {
        char hostname[256];
        if (gethostname(hostname, sizeof(hostname)) != 0) {
            return "localhost:" + std::to_string(listen_port_);
        }
        return std::string(hostname) + ":" + std::to_string(listen_port_);
    }

    void shutdown() {
        std::cerr << "[INFO] shutting down worker " << worker_id_ << "\n";

        if (server_) {
            server_->Shutdown();
            server_->Wait();
        }

        std::cerr << "[INFO] worker " << worker_id_ << " stopped\n";
    }

    std::string coordinator_addr_;
    std::string worker_id_;
    int capacity_;
    int listen_port_;
    std::map<std::string, std::string> labels_;
    std::map<std::string, std::unique_ptr<TaskHandler>> handlers_;

    std::atomic<int32_t> active_tasks_;
    std::atomic<bool> running_;
    std::mutex mu_;
    std::condition_variable cv_;

    std::unique_ptr<WorkerServicer> servicer_;
    std::unique_ptr<grpc::Server> server_;
};

// ---------------------------------------------------------------------------
// Worker public API — delegated to Impl via PIMPL.
// ---------------------------------------------------------------------------

Worker::Worker(const std::string& coordinator_addr,
               const std::string& worker_id)
    : impl_(std::make_unique<Impl>(coordinator_addr, worker_id)) {}

Worker::~Worker() = default;

void Worker::register_handler(std::unique_ptr<TaskHandler> handler) {
    impl_->register_handler(std::move(handler));
}

void Worker::set_capacity(int capacity) {
    impl_->set_capacity(capacity);
}

void Worker::set_labels(const std::map<std::string, std::string>& labels) {
    impl_->set_labels(labels);
}

void Worker::set_listen_port(int port) {
    impl_->set_listen_port(port);
}

const std::string& Worker::id() const noexcept {
    return impl_->id();
}

void Worker::start() {
    impl_->start();
}

void Worker::stop() {
    impl_->stop();
}

}  // namespace forge
